app/controllers/api/storage.php

```diff
@@ -24,9 +25,465 @@
+                    'array' => false,
+                    'filters' => [],
+                ]),
+            ], [
+                new Document([
+                    '$id' => '_key_bucket',
+                    'type' => Database::INDEX_KEY,
+                    'attributes' => ['bucketId'],
+                    'lengths' => [Database::LENGTH_KEY],
+                    'orders' => [Database::ORDER_ASC],
+                ]),
+                new Document([
+                    '$id' => '_key_search',
+                    'type' => Database::INDEX_FULLTEXT,
+                    'attributes' => ['search'],
+                    'lengths' => [2048],
+                    'orders' => [Database::ORDER_ASC],
+                ],),
+            ]);
+
+            $bucket = $dbForInternal->createDocument('buckets', new Document([
+                '$id' => $bucketId,
+                '$collection' => 'buckets',
+                'dateCreated' => \time(),
+                'dateUpdated' => \time(),
+            'name' => $name,
+                'permission' => $permission,
+                'maximumFileSize' => $maximumFileSize,
+                'allowedFileExtensions' => $allowedFileExtensions,
+                'enabled' => $enabled,
+                'adapter' => $adapter,
+                'encryption' => $encryption,
+                'antiVirus' => $antiVirus,
+                '$read' => $read ?? [],
+                '$write' => $write ?? [],
+                'search' => implode(' ', [$bucketId, $name]),
+            ]));
+        } catch (Duplicate $th) {
+            throw new Exception('Bucket already exists', 409);
+        }
+
+        $audits
+            ->setParam('event', 'storage.buckets.create')
+            ->setParam('resource', 'storage/buckets/' . $bucket->getId())
+            ->setParam('data', $bucket->getArrayCopy())
+        ;
+
+        $usage->setParam('storage.buckets.create', 1);
+
+        $response->setStatusCode(Response::STATUS_CODE_CREATED);
+        $response->dynamic($bucket, Response::MODEL_BUCKET);
+    });
+
+App::get('/v1/storage/buckets')
+    ->desc('List buckets')
+    ->groups(['api', 'storage'])
+    ->label('scope', 'buckets.read')
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'storage')
+    ->label('sdk.method', 'listBuckets')
+    ->label('sdk.description', '/docs/references/storage/list-buckets.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_BUCKET_LIST)
+    ->param('search', '', new Text(256), 'Search term to filter your list results. Max length: 256 chars.', true)
+    ->param('limit', 25, new Range(0, 100), 'Results limit value. By default will return maximum 25 results. Maximum of 100 results allowed per request.', true)
+    ->param('offset', 0, new Range(0, 2000), 'Results offset. The default value is 0. Use this param to manage pagination.', true)
```

Comment:
What do you think about setting this offset maximum to the count limit?
https://github.com/appwrite/appwrite/blob/252720dd2874e42988fff8f8c3619b0eeeebb0fa/app/init.php#L63

Reply:
I see different value in collections - for collections it seems to be 40000 and for documents it seems to be 9000000 ? What should I do for buckets and files? and how are we deciding this?